File: go/pkg/models/document_location.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type LocationType string

const (
	LocationTypeCity      LocationType = "city"
	LocationTypeCountry   LocationType = "country"
	LocationTypeContinent LocationType = "continent"
)

type DocumentLocation struct {
	ID               uuid.UUID    `gorm:"type:uuid;primary_key;" json:"id"`
	CreatedAt        time.Time    `json:"createdAt"`
	UpdatedAt        time.Time    `json:"updatedAt"`
	DocumentID       uuid.UUID    `gorm:"type:uuid;not null" json:"documentId"`
	Document         Document     `json:"document" gorm:"foreignKey:DocumentID"`
	PlaceID          string       `gorm:"type:varchar;not null" json:"placeId"`
	Name             string       `gorm:"type:varchar;not null" json:"name"`
	FormattedAddress string       `gorm:"type:text;not null" json:"formattedAddress"`
	Latitude         float64      `gorm:"type:double precision;not null" json:"latitude"`
	Longitude        float64      `gorm:"type:double precision;not null" json:"longitude"`
	LocationType     LocationType `gorm:"type:varchar;not null" json:"locationType"`
}

func (DocumentLocation) TableName() string {
	return "document_locations"
}

File: dart/travel_angels/lib/models/document_location.dart
import 'package:json_annotation/json_annotation.dart';

part 'document_location.g.dart';

enum LocationType {
  @JsonValue('city')
  city,
  @JsonValue('country')
  country,
  @JsonValue('continent')
  continent,
}

@JsonSerializable()
class DocumentLocation {
  @JsonKey(name: 'id', fromJson: _idFromJson)
  final String id;

  @JsonKey(name: 'documentId', fromJson: _idFromJson)
  final String documentId;

  @JsonKey(name: 'placeId')
  final String placeId;

  @JsonKey(name: 'name')
  final String name;

  @JsonKey(name: 'formattedAddress')
  final String formattedAddress;

  @JsonKey(name: 'latitude')
  final double latitude;

  @JsonKey(name: 'longitude')
  final double longitude;

  @JsonKey(name: 'locationType', fromJson: _locationTypeFromJson)
  final LocationType locationType;

  DocumentLocation({
    required this.id,
    required this.documentId,
    required this.placeId,
    required this.name,
    required this.formattedAddress,
    required this.latitude,
    required this.longitude,
    required this.locationType,
  });

  factory DocumentLocation.fromJson(Map<String, dynamic> json) =>
      _$DocumentLocationFromJson(json);

  Map<String, dynamic> toJson() => _$DocumentLocationToJson(this);

  static String _idFromJson(dynamic json) {
    if (json == null) return '';
    return json.toString();
  }

  static LocationType _locationTypeFromJson(dynamic json) {
    if (json == null) return LocationType.city;
    final str = json.toString().toLowerCase();
    switch (str) {
      case 'country':
        return LocationType.country;
      case 'continent':
        return LocationType.continent;
      case 'city':
      default:
        return LocationType.city;
    }
  }
}

